Description: Fix test errors caused by hardcoded sizes
 Cherry-picked as of commit 3260252708907eb0405d3250e72779ca790a5eea
Last-Update: 2024-01-31
---
This patch header follows DEP-3: http://dep.debian.net/deps/dep3/
--- a/src/pam/sys.rs
+++ b/src/pam/sys.rs
@@ -1,86 +1,53 @@
-/* automatically generated by rust-bindgen 0.66.1 */
+/* automatically generated by rust-bindgen 0.66.1, minified by cargo-minify, edited to be portable */
 
-pub const PAM_SUCCESS: u32 = 0;
-pub const PAM_OPEN_ERR: u32 = 1;
-pub const PAM_SYMBOL_ERR: u32 = 2;
-pub const PAM_SERVICE_ERR: u32 = 3;
-pub const PAM_SYSTEM_ERR: u32 = 4;
-pub const PAM_BUF_ERR: u32 = 5;
-pub const PAM_PERM_DENIED: u32 = 6;
-pub const PAM_AUTH_ERR: u32 = 7;
-pub const PAM_CRED_INSUFFICIENT: u32 = 8;
-pub const PAM_AUTHINFO_UNAVAIL: u32 = 9;
-pub const PAM_USER_UNKNOWN: u32 = 10;
-pub const PAM_MAXTRIES: u32 = 11;
-pub const PAM_NEW_AUTHTOK_REQD: u32 = 12;
-pub const PAM_ACCT_EXPIRED: u32 = 13;
-pub const PAM_SESSION_ERR: u32 = 14;
-pub const PAM_CRED_UNAVAIL: u32 = 15;
-pub const PAM_CRED_EXPIRED: u32 = 16;
-pub const PAM_CRED_ERR: u32 = 17;
-pub const PAM_NO_MODULE_DATA: u32 = 18;
-pub const PAM_CONV_ERR: u32 = 19;
-pub const PAM_AUTHTOK_ERR: u32 = 20;
-pub const PAM_AUTHTOK_RECOVERY_ERR: u32 = 21;
-pub const PAM_AUTHTOK_LOCK_BUSY: u32 = 22;
-pub const PAM_AUTHTOK_DISABLE_AGING: u32 = 23;
-pub const PAM_TRY_AGAIN: u32 = 24;
-pub const PAM_IGNORE: u32 = 25;
-pub const PAM_ABORT: u32 = 26;
-pub const PAM_AUTHTOK_EXPIRED: u32 = 27;
-pub const PAM_MODULE_UNKNOWN: u32 = 28;
-pub const PAM_BAD_ITEM: u32 = 29;
-pub const PAM_CONV_AGAIN: u32 = 30;
-pub const PAM_INCOMPLETE: u32 = 31;
-pub const PAM_SILENT: u32 = 32768;
-pub const PAM_DISALLOW_NULL_AUTHTOK: u32 = 1;
-pub const PAM_ESTABLISH_CRED: u32 = 2;
-pub const PAM_DELETE_CRED: u32 = 4;
-pub const PAM_REINITIALIZE_CRED: u32 = 8;
-pub const PAM_REFRESH_CRED: u32 = 16;
-pub const PAM_CHANGE_EXPIRED_AUTHTOK: u32 = 32;
-pub const PAM_SERVICE: u32 = 1;
-pub const PAM_USER: u32 = 2;
-pub const PAM_TTY: u32 = 3;
-pub const PAM_RHOST: u32 = 4;
-pub const PAM_CONV: u32 = 5;
-pub const PAM_AUTHTOK: u32 = 6;
-pub const PAM_OLDAUTHTOK: u32 = 7;
-pub const PAM_RUSER: u32 = 8;
-pub const PAM_USER_PROMPT: u32 = 9;
-pub const PAM_FAIL_DELAY: u32 = 10;
-pub const PAM_XDISPLAY: u32 = 11;
-pub const PAM_XAUTHDATA: u32 = 12;
-pub const PAM_AUTHTOK_TYPE: u32 = 13;
-pub const PAM_DATA_SILENT: u32 = 1073741824;
-pub const PAM_PROMPT_ECHO_OFF: u32 = 1;
-pub const PAM_PROMPT_ECHO_ON: u32 = 2;
-pub const PAM_ERROR_MSG: u32 = 3;
-pub const PAM_TEXT_INFO: u32 = 4;
-pub const PAM_RADIO_TYPE: u32 = 5;
-pub const PAM_BINARY_PROMPT: u32 = 7;
-pub const PAM_MAX_NUM_MSG: u32 = 32;
-pub const PAM_MAX_MSG_SIZE: u32 = 512;
-pub const PAM_MAX_RESP_SIZE: u32 = 512;
-pub const PAM_AUTHTOK_RECOVER_ERR: u32 = 21;
-pub const PAM_BP_MAX_LENGTH: u32 = 131072;
-pub const PAM_BPC_FALSE: u32 = 0;
-pub const PAM_BPC_TRUE: u32 = 1;
-pub const PAM_BPC_OK: u32 = 1;
-pub const PAM_BPC_SELECT: u32 = 2;
-pub const PAM_BPC_DONE: u32 = 3;
-pub const PAM_BPC_FAIL: u32 = 4;
-pub const PAM_BPC_GETENV: u32 = 65;
-pub const PAM_BPC_PUTENV: u32 = 66;
-pub const PAM_BPC_TEXT: u32 = 67;
-pub const PAM_BPC_ERROR: u32 = 68;
-pub const PAM_BPC_PROMPT: u32 = 69;
-pub const PAM_BPC_PASS: u32 = 70;
-pub const PAM_PRELIM_CHECK: u32 = 16384;
-pub const PAM_UPDATE_AUTHTOK: u32 = 8192;
-pub const PAM_DATA_REPLACE: u32 = 536870912;
-pub const PAM_MODUTIL_NGROUPS: u32 = 64;
-pub type pam_handle_t = u8;
+// NOTE: the tests below test the assumptions about the padding that a C compiler will use on the
+// above structs; if these assumptions are incorrect, the tests will fail, but most likely the
+// code will still be correct.
+
+pub const PAM_SUCCESS: libc::c_int = 0;
+pub const PAM_OPEN_ERR: libc::c_int = 1;
+pub const PAM_SYMBOL_ERR: libc::c_int = 2;
+pub const PAM_SERVICE_ERR: libc::c_int = 3;
+pub const PAM_SYSTEM_ERR: libc::c_int = 4;
+pub const PAM_BUF_ERR: libc::c_int = 5;
+pub const PAM_PERM_DENIED: libc::c_int = 6;
+pub const PAM_AUTH_ERR: libc::c_int = 7;
+pub const PAM_CRED_INSUFFICIENT: libc::c_int = 8;
+pub const PAM_AUTHINFO_UNAVAIL: libc::c_int = 9;
+pub const PAM_USER_UNKNOWN: libc::c_int = 10;
+pub const PAM_MAXTRIES: libc::c_int = 11;
+pub const PAM_NEW_AUTHTOK_REQD: libc::c_int = 12;
+pub const PAM_ACCT_EXPIRED: libc::c_int = 13;
+pub const PAM_SESSION_ERR: libc::c_int = 14;
+pub const PAM_CRED_UNAVAIL: libc::c_int = 15;
+pub const PAM_CRED_EXPIRED: libc::c_int = 16;
+pub const PAM_CRED_ERR: libc::c_int = 17;
+pub const PAM_NO_MODULE_DATA: libc::c_int = 18;
+pub const PAM_CONV_ERR: libc::c_int = 19;
+pub const PAM_AUTHTOK_ERR: libc::c_int = 20;
+pub const PAM_AUTHTOK_RECOVERY_ERR: libc::c_int = 21;
+pub const PAM_AUTHTOK_LOCK_BUSY: libc::c_int = 22;
+pub const PAM_AUTHTOK_DISABLE_AGING: libc::c_int = 23;
+pub const PAM_TRY_AGAIN: libc::c_int = 24;
+pub const PAM_IGNORE: libc::c_int = 25;
+pub const PAM_ABORT: libc::c_int = 26;
+pub const PAM_AUTHTOK_EXPIRED: libc::c_int = 27;
+pub const PAM_MODULE_UNKNOWN: libc::c_int = 28;
+pub const PAM_BAD_ITEM: libc::c_int = 29;
+pub const PAM_SILENT: libc::c_int = 32768;
+pub const PAM_DISALLOW_NULL_AUTHTOK: libc::c_int = 1;
+pub const PAM_REINITIALIZE_CRED: libc::c_int = 8;
+pub const PAM_CHANGE_EXPIRED_AUTHTOK: libc::c_int = 32;
+pub const PAM_USER: libc::c_int = 2;
+pub const PAM_TTY: libc::c_int = 3;
+pub const PAM_RUSER: libc::c_int = 8;
+pub const PAM_DATA_SILENT: libc::c_int = 1073741824;
+pub const PAM_PROMPT_ECHO_OFF: libc::c_int = 1;
+pub const PAM_PROMPT_ECHO_ON: libc::c_int = 2;
+pub const PAM_ERROR_MSG: libc::c_int = 3;
+pub const PAM_TEXT_INFO: libc::c_int = 4;
+pub const PAM_MAX_RESP_SIZE: libc::c_int = 512;
+pub type pam_handle_t = libc::c_void;
 extern "C" {
     pub fn pam_set_item(
         pamh: *mut pam_handle_t,
@@ -99,17 +66,8 @@
     pub fn pam_strerror(pamh: *mut pam_handle_t, errnum: libc::c_int) -> *const libc::c_char;
 }
 extern "C" {
-    pub fn pam_putenv(pamh: *mut pam_handle_t, name_value: *const libc::c_char) -> libc::c_int;
-}
-extern "C" {
-    pub fn pam_getenv(pamh: *mut pam_handle_t, name: *const libc::c_char) -> *const libc::c_char;
-}
-extern "C" {
     pub fn pam_getenvlist(pamh: *mut pam_handle_t) -> *mut *mut libc::c_char;
 }
-extern "C" {
-    pub fn pam_fail_delay(pamh: *mut pam_handle_t, musec_delay: libc::c_uint) -> libc::c_int;
-}
 #[repr(C)]
 #[derive(Debug, Copy, Clone)]
 pub struct pam_message {
@@ -121,18 +79,14 @@
     const UNINIT: ::std::mem::MaybeUninit<pam_message> = ::std::mem::MaybeUninit::uninit();
     let ptr = UNINIT.as_ptr();
     assert_eq!(
-        ::std::mem::size_of::<pam_message>(),
-        16usize,
-        concat!("Size of: ", stringify!(pam_message))
-    );
-    assert_eq!(
         ::std::mem::align_of::<pam_message>(),
-        8usize,
+        ::std::mem::align_of::<*mut libc::c_void>(),
         concat!("Alignment of ", stringify!(pam_message))
     );
+    let mut offset: usize = 0;
     assert_eq!(
         unsafe { ::std::ptr::addr_of!((*ptr).msg_style) as usize - ptr as usize },
-        0usize,
+        offset,
         concat!(
             "Offset of field: ",
             stringify!(pam_message),
@@ -140,9 +94,10 @@
             stringify!(msg_style)
         )
     );
+    offset = aligned_offset::<*const libc::c_char>(offset + ::std::mem::size_of::<libc::c_int>());
     assert_eq!(
         unsafe { ::std::ptr::addr_of!((*ptr).msg) as usize - ptr as usize },
-        8usize,
+        offset,
         concat!(
             "Offset of field: ",
             stringify!(pam_message),
@@ -150,6 +105,14 @@
             stringify!(msg)
         )
     );
+    offset = aligned_offset::<*const libc::c_void>(
+        offset + ::std::mem::size_of::<*const libc::c_char>(),
+    );
+    assert_eq!(
+        ::std::mem::size_of::<pam_message>(),
+        offset,
+        concat!("Size of: ", stringify!(pam_message))
+    );
 }
 #[repr(C)]
 #[derive(Debug, Copy, Clone)]
@@ -162,18 +125,14 @@
     const UNINIT: ::std::mem::MaybeUninit<pam_response> = ::std::mem::MaybeUninit::uninit();
     let ptr = UNINIT.as_ptr();
     assert_eq!(
-        ::std::mem::size_of::<pam_response>(),
-        16usize,
-        concat!("Size of: ", stringify!(pam_response))
-    );
-    assert_eq!(
         ::std::mem::align_of::<pam_response>(),
-        8usize,
+        ::std::mem::align_of::<*mut libc::c_char>(),
         concat!("Alignment of ", stringify!(pam_response))
     );
+    let mut offset: usize = 0;
     assert_eq!(
         unsafe { ::std::ptr::addr_of!((*ptr).resp) as usize - ptr as usize },
-        0usize,
+        offset,
         concat!(
             "Offset of field: ",
             stringify!(pam_response),
@@ -181,9 +140,10 @@
             stringify!(resp)
         )
     );
+    offset = aligned_offset::<libc::c_int>(offset + ::std::mem::size_of::<*mut libc::c_char>());
     assert_eq!(
         unsafe { ::std::ptr::addr_of!((*ptr).resp_retcode) as usize - ptr as usize },
-        8usize,
+        offset,
         concat!(
             "Offset of field: ",
             stringify!(pam_response),
@@ -191,6 +151,12 @@
             stringify!(resp_retcode)
         )
     );
+    offset = aligned_offset::<*mut libc::c_void>(offset + ::std::mem::size_of::<libc::c_int>());
+    assert_eq!(
+        ::std::mem::size_of::<pam_response>(),
+        offset,
+        concat!("Size of: ", stringify!(pam_response))
+    );
 }
 #[repr(C)]
 #[derive(Debug, Copy, Clone)]
@@ -210,18 +176,14 @@
     const UNINIT: ::std::mem::MaybeUninit<pam_conv> = ::std::mem::MaybeUninit::uninit();
     let ptr = UNINIT.as_ptr();
     assert_eq!(
-        ::std::mem::size_of::<pam_conv>(),
-        16usize,
-        concat!("Size of: ", stringify!(pam_conv))
-    );
-    assert_eq!(
         ::std::mem::align_of::<pam_conv>(),
-        8usize,
+        ::std::mem::align_of::<*mut libc::c_void>(),
         concat!("Alignment of ", stringify!(pam_conv))
     );
+    let mut offset: usize = 0;
     assert_eq!(
         unsafe { ::std::ptr::addr_of!((*ptr).conv) as usize - ptr as usize },
-        0usize,
+        offset,
         concat!(
             "Offset of field: ",
             stringify!(pam_conv),
@@ -229,9 +191,11 @@
             stringify!(conv)
         )
     );
+    offset =
+        aligned_offset::<*mut libc::c_void>(offset + ::std::mem::size_of::<*mut libc::c_void>());
     assert_eq!(
         unsafe { ::std::ptr::addr_of!((*ptr).appdata_ptr) as usize - ptr as usize },
-        8usize,
+        offset,
         concat!(
             "Offset of field: ",
             stringify!(pam_conv),
@@ -239,6 +203,13 @@
             stringify!(appdata_ptr)
         )
     );
+    offset =
+        aligned_offset::<*mut libc::c_void>(offset + ::std::mem::size_of::<*mut libc::c_void>());
+    assert_eq!(
+        ::std::mem::size_of::<pam_conv>(),
+        offset,
+        concat!("Size of: ", stringify!(pam_conv))
+    );
 }
 extern "C" {
     pub fn pam_start(
@@ -249,15 +220,6 @@
     ) -> libc::c_int;
 }
 extern "C" {
-    pub fn pam_start_confdir(
-        service_name: *const libc::c_char,
-        user: *const libc::c_char,
-        pam_conversation: *const pam_conv,
-        confdir: *const libc::c_char,
-        pamh: *mut *mut pam_handle_t,
-    ) -> libc::c_int;
-}
-extern "C" {
     pub fn pam_end(pamh: *mut pam_handle_t, pam_status: libc::c_int) -> libc::c_int;
 }
 extern "C" {
@@ -282,156 +244,6 @@
 pub type __gid_t = libc::c_uint;
 pub type gid_t = __gid_t;
 pub type uid_t = __uid_t;
-pub type va_list = __builtin_va_list;
-extern "C" {
-    pub fn pam_vsyslog(
-        pamh: *const pam_handle_t,
-        priority: libc::c_int,
-        fmt: *const libc::c_char,
-        args: *mut __va_list_tag,
-    );
-}
-extern "C" {
-    pub fn pam_syslog(
-        pamh: *const pam_handle_t,
-        priority: libc::c_int,
-        fmt: *const libc::c_char,
-        ...
-    );
-}
-extern "C" {
-    pub fn pam_vprompt(
-        pamh: *mut pam_handle_t,
-        style: libc::c_int,
-        response: *mut *mut libc::c_char,
-        fmt: *const libc::c_char,
-        args: *mut __va_list_tag,
-    ) -> libc::c_int;
-}
-extern "C" {
-    pub fn pam_prompt(
-        pamh: *mut pam_handle_t,
-        style: libc::c_int,
-        response: *mut *mut libc::c_char,
-        fmt: *const libc::c_char,
-        ...
-    ) -> libc::c_int;
-}
-extern "C" {
-    pub fn pam_get_authtok(
-        pamh: *mut pam_handle_t,
-        item: libc::c_int,
-        authtok: *mut *const libc::c_char,
-        prompt: *const libc::c_char,
-    ) -> libc::c_int;
-}
-extern "C" {
-    pub fn pam_get_authtok_noverify(
-        pamh: *mut pam_handle_t,
-        authtok: *mut *const libc::c_char,
-        prompt: *const libc::c_char,
-    ) -> libc::c_int;
-}
-extern "C" {
-    pub fn pam_get_authtok_verify(
-        pamh: *mut pam_handle_t,
-        authtok: *mut *const libc::c_char,
-        prompt: *const libc::c_char,
-    ) -> libc::c_int;
-}
-extern "C" {
-    pub fn pam_misc_paste_env(
-        pamh: *mut pam_handle_t,
-        user_env: *const *const libc::c_char,
-    ) -> libc::c_int;
-}
-extern "C" {
-    pub fn pam_misc_drop_env(env: *mut *mut libc::c_char) -> *mut *mut libc::c_char;
-}
-extern "C" {
-    pub fn pam_misc_setenv(
-        pamh: *mut pam_handle_t,
-        name: *const libc::c_char,
-        value: *const libc::c_char,
-        readonly: libc::c_int,
-    ) -> libc::c_int;
-}
-extern "C" {
-    pub fn pam_set_data(
-        pamh: *mut pam_handle_t,
-        module_data_name: *const libc::c_char,
-        data: *mut libc::c_void,
-        cleanup: ::std::option::Option<
-            unsafe extern "C" fn(
-                pamh: *mut pam_handle_t,
-                data: *mut libc::c_void,
-                error_status: libc::c_int,
-            ),
-        >,
-    ) -> libc::c_int;
-}
-extern "C" {
-    pub fn pam_get_data(
-        pamh: *const pam_handle_t,
-        module_data_name: *const libc::c_char,
-        data: *mut *const libc::c_void,
-    ) -> libc::c_int;
-}
-extern "C" {
-    pub fn pam_get_user(
-        pamh: *mut pam_handle_t,
-        user: *mut *const libc::c_char,
-        prompt: *const libc::c_char,
-    ) -> libc::c_int;
-}
-extern "C" {
-    pub fn pam_sm_authenticate(
-        pamh: *mut pam_handle_t,
-        flags: libc::c_int,
-        argc: libc::c_int,
-        argv: *mut *const libc::c_char,
-    ) -> libc::c_int;
-}
-extern "C" {
-    pub fn pam_sm_setcred(
-        pamh: *mut pam_handle_t,
-        flags: libc::c_int,
-        argc: libc::c_int,
-        argv: *mut *const libc::c_char,
-    ) -> libc::c_int;
-}
-extern "C" {
-    pub fn pam_sm_acct_mgmt(
-        pamh: *mut pam_handle_t,
-        flags: libc::c_int,
-        argc: libc::c_int,
-        argv: *mut *const libc::c_char,
-    ) -> libc::c_int;
-}
-extern "C" {
-    pub fn pam_sm_open_session(
-        pamh: *mut pam_handle_t,
-        flags: libc::c_int,
-        argc: libc::c_int,
-        argv: *mut *const libc::c_char,
-    ) -> libc::c_int;
-}
-extern "C" {
-    pub fn pam_sm_close_session(
-        pamh: *mut pam_handle_t,
-        flags: libc::c_int,
-        argc: libc::c_int,
-        argv: *mut *const libc::c_char,
-    ) -> libc::c_int;
-}
-extern "C" {
-    pub fn pam_sm_chauthtok(
-        pamh: *mut pam_handle_t,
-        flags: libc::c_int,
-        argc: libc::c_int,
-        argv: *mut *const libc::c_char,
-    ) -> libc::c_int;
-}
 #[repr(C)]
 #[derive(Debug, Copy, Clone)]
 pub struct passwd {
@@ -448,18 +260,14 @@
     const UNINIT: ::std::mem::MaybeUninit<passwd> = ::std::mem::MaybeUninit::uninit();
     let ptr = UNINIT.as_ptr();
     assert_eq!(
-        ::std::mem::size_of::<passwd>(),
-        48usize,
-        concat!("Size of: ", stringify!(passwd))
-    );
-    assert_eq!(
         ::std::mem::align_of::<passwd>(),
-        8usize,
+        ::std::mem::align_of::<*mut libc::c_char>(),
         concat!("Alignment of ", stringify!(passwd))
     );
+    let mut offset: usize = 0;
     assert_eq!(
         unsafe { ::std::ptr::addr_of!((*ptr).pw_name) as usize - ptr as usize },
-        0usize,
+        offset,
         concat!(
             "Offset of field: ",
             stringify!(passwd),
@@ -467,9 +275,11 @@
             stringify!(pw_name)
         )
     );
+    offset =
+        aligned_offset::<*mut libc::c_char>(offset + ::std::mem::size_of::<*mut libc::c_char>());
     assert_eq!(
         unsafe { ::std::ptr::addr_of!((*ptr).pw_passwd) as usize - ptr as usize },
-        8usize,
+        offset,
         concat!(
             "Offset of field: ",
             stringify!(passwd),
@@ -477,9 +287,10 @@
             stringify!(pw_passwd)
         )
     );
+    offset = aligned_offset::<__uid_t>(offset + ::std::mem::size_of::<*mut libc::c_char>());
     assert_eq!(
         unsafe { ::std::ptr::addr_of!((*ptr).pw_uid) as usize - ptr as usize },
-        16usize,
+        offset,
         concat!(
             "Offset of field: ",
             stringify!(passwd),
@@ -487,9 +298,10 @@
             stringify!(pw_uid)
         )
     );
+    offset = aligned_offset::<__gid_t>(offset + ::std::mem::size_of::<__uid_t>());
     assert_eq!(
         unsafe { ::std::ptr::addr_of!((*ptr).pw_gid) as usize - ptr as usize },
-        20usize,
+        offset,
         concat!(
             "Offset of field: ",
             stringify!(passwd),
@@ -497,9 +309,10 @@
             stringify!(pw_gid)
         )
     );
+    offset = aligned_offset::<*mut libc::c_char>(offset + ::std::mem::size_of::<__gid_t>());
     assert_eq!(
         unsafe { ::std::ptr::addr_of!((*ptr).pw_gecos) as usize - ptr as usize },
-        24usize,
+        offset,
         concat!(
             "Offset of field: ",
             stringify!(passwd),
@@ -507,9 +320,11 @@
             stringify!(pw_gecos)
         )
     );
+    offset =
+        aligned_offset::<*mut libc::c_char>(offset + ::std::mem::size_of::<*mut libc::c_char>());
     assert_eq!(
         unsafe { ::std::ptr::addr_of!((*ptr).pw_dir) as usize - ptr as usize },
-        32usize,
+        offset,
         concat!(
             "Offset of field: ",
             stringify!(passwd),
@@ -517,9 +332,11 @@
             stringify!(pw_dir)
         )
     );
+    offset =
+        aligned_offset::<*mut libc::c_char>(offset + ::std::mem::size_of::<*mut libc::c_char>());
     assert_eq!(
         unsafe { ::std::ptr::addr_of!((*ptr).pw_shell) as usize - ptr as usize },
-        40usize,
+        offset,
         concat!(
             "Offset of field: ",
             stringify!(passwd),
@@ -527,6 +344,13 @@
             stringify!(pw_shell)
         )
     );
+    offset =
+        aligned_offset::<*mut libc::c_void>(offset + ::std::mem::size_of::<*mut libc::c_char>());
+    assert_eq!(
+        ::std::mem::size_of::<passwd>(),
+        offset,
+        concat!("Size of: ", stringify!(passwd))
+    );
 }
 #[repr(C)]
 #[derive(Debug, Copy, Clone)]
@@ -541,18 +365,14 @@
     const UNINIT: ::std::mem::MaybeUninit<group> = ::std::mem::MaybeUninit::uninit();
     let ptr = UNINIT.as_ptr();
     assert_eq!(
-        ::std::mem::size_of::<group>(),
-        32usize,
-        concat!("Size of: ", stringify!(group))
-    );
-    assert_eq!(
         ::std::mem::align_of::<group>(),
-        8usize,
+        ::std::mem::align_of::<*mut libc::c_char>(),
         concat!("Alignment of ", stringify!(group))
     );
+    let mut offset: usize = 0;
     assert_eq!(
         unsafe { ::std::ptr::addr_of!((*ptr).gr_name) as usize - ptr as usize },
-        0usize,
+        offset,
         concat!(
             "Offset of field: ",
             stringify!(group),
@@ -560,9 +380,11 @@
             stringify!(gr_name)
         )
     );
+    offset =
+        aligned_offset::<*mut libc::c_char>(offset + ::std::mem::size_of::<*mut libc::c_char>());
     assert_eq!(
         unsafe { ::std::ptr::addr_of!((*ptr).gr_passwd) as usize - ptr as usize },
-        8usize,
+        offset,
         concat!(
             "Offset of field: ",
             stringify!(group),
@@ -570,9 +392,10 @@
             stringify!(gr_passwd)
         )
     );
+    offset = aligned_offset::<__gid_t>(offset + ::std::mem::size_of::<*mut libc::c_char>());
     assert_eq!(
         unsafe { ::std::ptr::addr_of!((*ptr).gr_gid) as usize - ptr as usize },
-        16usize,
+        offset,
         concat!(
             "Offset of field: ",
             stringify!(group),
@@ -580,9 +403,10 @@
             stringify!(gr_gid)
         )
     );
+    offset = aligned_offset::<*mut libc::c_char>(offset + ::std::mem::size_of::<__gid_t>());
     assert_eq!(
         unsafe { ::std::ptr::addr_of!((*ptr).gr_mem) as usize - ptr as usize },
-        24usize,
+        offset,
         concat!(
             "Offset of field: ",
             stringify!(group),
@@ -590,6 +414,13 @@
             stringify!(gr_mem)
         )
     );
+    offset =
+        aligned_offset::<*mut libc::c_void>(offset + ::std::mem::size_of::<*mut libc::c_char>());
+    assert_eq!(
+        ::std::mem::size_of::<group>(),
+        offset,
+        concat!("Size of: ", stringify!(group))
+    );
 }
 #[repr(C)]
 #[derive(Debug, Copy, Clone)]
@@ -609,18 +440,14 @@
     const UNINIT: ::std::mem::MaybeUninit<spwd> = ::std::mem::MaybeUninit::uninit();
     let ptr = UNINIT.as_ptr();
     assert_eq!(
-        ::std::mem::size_of::<spwd>(),
-        72usize,
-        concat!("Size of: ", stringify!(spwd))
-    );
-    assert_eq!(
         ::std::mem::align_of::<spwd>(),
-        8usize,
+        ::std::mem::align_of::<*mut libc::c_char>(),
         concat!("Alignment of ", stringify!(spwd))
     );
+    let mut offset: usize = 0;
     assert_eq!(
         unsafe { ::std::ptr::addr_of!((*ptr).sp_namp) as usize - ptr as usize },
-        0usize,
+        offset,
         concat!(
             "Offset of field: ",
             stringify!(spwd),
@@ -628,9 +455,11 @@
             stringify!(sp_namp)
         )
     );
+    offset =
+        aligned_offset::<*mut libc::c_char>(offset + ::std::mem::size_of::<*mut libc::c_char>());
     assert_eq!(
         unsafe { ::std::ptr::addr_of!((*ptr).sp_pwdp) as usize - ptr as usize },
-        8usize,
+        offset,
         concat!(
             "Offset of field: ",
             stringify!(spwd),
@@ -638,9 +467,10 @@
             stringify!(sp_pwdp)
         )
     );
+    offset = aligned_offset::<libc::c_long>(offset + ::std::mem::size_of::<*mut libc::c_char>());
     assert_eq!(
         unsafe { ::std::ptr::addr_of!((*ptr).sp_lstchg) as usize - ptr as usize },
-        16usize,
+        offset,
         concat!(
             "Offset of field: ",
             stringify!(spwd),
@@ -648,9 +478,10 @@
             stringify!(sp_lstchg)
         )
     );
+    offset = aligned_offset::<libc::c_long>(offset + ::std::mem::size_of::<libc::c_long>());
     assert_eq!(
         unsafe { ::std::ptr::addr_of!((*ptr).sp_min) as usize - ptr as usize },
-        24usize,
+        offset,
         concat!(
             "Offset of field: ",
             stringify!(spwd),
@@ -658,9 +489,10 @@
             stringify!(sp_min)
         )
     );
+    offset = aligned_offset::<libc::c_long>(offset + ::std::mem::size_of::<libc::c_long>());
     assert_eq!(
         unsafe { ::std::ptr::addr_of!((*ptr).sp_max) as usize - ptr as usize },
-        32usize,
+        offset,
         concat!(
             "Offset of field: ",
             stringify!(spwd),
@@ -668,9 +500,10 @@
             stringify!(sp_max)
         )
     );
+    offset = aligned_offset::<libc::c_long>(offset + ::std::mem::size_of::<libc::c_long>());
     assert_eq!(
         unsafe { ::std::ptr::addr_of!((*ptr).sp_warn) as usize - ptr as usize },
-        40usize,
+        offset,
         concat!(
             "Offset of field: ",
             stringify!(spwd),
@@ -678,9 +511,10 @@
             stringify!(sp_warn)
         )
     );
+    offset = aligned_offset::<libc::c_long>(offset + ::std::mem::size_of::<libc::c_long>());
     assert_eq!(
         unsafe { ::std::ptr::addr_of!((*ptr).sp_inact) as usize - ptr as usize },
-        48usize,
+        offset,
         concat!(
             "Offset of field: ",
             stringify!(spwd),
@@ -688,9 +522,10 @@
             stringify!(sp_inact)
         )
     );
+    offset = aligned_offset::<libc::c_long>(offset + ::std::mem::size_of::<libc::c_long>());
     assert_eq!(
         unsafe { ::std::ptr::addr_of!((*ptr).sp_expire) as usize - ptr as usize },
-        56usize,
+        offset,
         concat!(
             "Offset of field: ",
             stringify!(spwd),
@@ -698,9 +533,10 @@
             stringify!(sp_expire)
         )
     );
+    offset = aligned_offset::<libc::c_long>(offset + ::std::mem::size_of::<libc::c_long>());
     assert_eq!(
         unsafe { ::std::ptr::addr_of!((*ptr).sp_flag) as usize - ptr as usize },
-        64usize,
+        offset,
         concat!(
             "Offset of field: ",
             stringify!(spwd),
@@ -708,74 +544,12 @@
             stringify!(sp_flag)
         )
     );
-}
-extern "C" {
-    pub fn pam_modutil_getpwnam(pamh: *mut pam_handle_t, user: *const libc::c_char) -> *mut passwd;
-}
-extern "C" {
-    pub fn pam_modutil_getpwuid(pamh: *mut pam_handle_t, uid: uid_t) -> *mut passwd;
-}
-extern "C" {
-    pub fn pam_modutil_getgrnam(pamh: *mut pam_handle_t, group: *const libc::c_char) -> *mut group;
-}
-extern "C" {
-    pub fn pam_modutil_getgrgid(pamh: *mut pam_handle_t, gid: gid_t) -> *mut group;
-}
-extern "C" {
-    pub fn pam_modutil_getspnam(pamh: *mut pam_handle_t, user: *const libc::c_char) -> *mut spwd;
-}
-extern "C" {
-    pub fn pam_modutil_user_in_group_nam_nam(
-        pamh: *mut pam_handle_t,
-        user: *const libc::c_char,
-        group: *const libc::c_char,
-    ) -> libc::c_int;
-}
-extern "C" {
-    pub fn pam_modutil_user_in_group_nam_gid(
-        pamh: *mut pam_handle_t,
-        user: *const libc::c_char,
-        group: gid_t,
-    ) -> libc::c_int;
-}
-extern "C" {
-    pub fn pam_modutil_user_in_group_uid_nam(
-        pamh: *mut pam_handle_t,
-        user: uid_t,
-        group: *const libc::c_char,
-    ) -> libc::c_int;
-}
-extern "C" {
-    pub fn pam_modutil_user_in_group_uid_gid(
-        pamh: *mut pam_handle_t,
-        user: uid_t,
-        group: gid_t,
-    ) -> libc::c_int;
-}
-extern "C" {
-    pub fn pam_modutil_getlogin(pamh: *mut pam_handle_t) -> *const libc::c_char;
-}
-extern "C" {
-    pub fn pam_modutil_read(
-        fd: libc::c_int,
-        buffer: *mut libc::c_char,
-        count: libc::c_int,
-    ) -> libc::c_int;
-}
-extern "C" {
-    pub fn pam_modutil_write(
-        fd: libc::c_int,
-        buffer: *const libc::c_char,
-        count: libc::c_int,
-    ) -> libc::c_int;
-}
-extern "C" {
-    pub fn pam_modutil_audit_write(
-        pamh: *mut pam_handle_t,
-        type_: libc::c_int,
-        message: *const libc::c_char,
-        retval: libc::c_int,
-    ) -> libc::c_int;
+    offset = aligned_offset::<*mut libc::c_void>(offset + ::std::mem::size_of::<libc::c_long>());
+    assert_eq!(
+        ::std::mem::size_of::<spwd>(),
+        offset,
+        concat!("Size of: ", stringify!(spwd))
+    );
 }
 #[repr(C)]
 #[derive(Debug, Copy, Clone)]
@@ -792,18 +566,14 @@
     const UNINIT: ::std::mem::MaybeUninit<pam_modutil_privs> = ::std::mem::MaybeUninit::uninit();
     let ptr = UNINIT.as_ptr();
     assert_eq!(
-        ::std::mem::size_of::<pam_modutil_privs>(),
-        32usize,
-        concat!("Size of: ", stringify!(pam_modutil_privs))
-    );
-    assert_eq!(
         ::std::mem::align_of::<pam_modutil_privs>(),
-        8usize,
+        ::std::mem::align_of::<*mut gid_t>(),
         concat!("Alignment of ", stringify!(pam_modutil_privs))
     );
+    let mut offset: usize = 0;
     assert_eq!(
         unsafe { ::std::ptr::addr_of!((*ptr).grplist) as usize - ptr as usize },
-        0usize,
+        offset,
         concat!(
             "Offset of field: ",
             stringify!(pam_modutil_privs),
@@ -811,9 +581,10 @@
             stringify!(grplist)
         )
     );
+    offset = aligned_offset::<libc::c_int>(offset + ::std::mem::size_of::<*mut gid_t>());
     assert_eq!(
         unsafe { ::std::ptr::addr_of!((*ptr).number_of_groups) as usize - ptr as usize },
-        8usize,
+        offset,
         concat!(
             "Offset of field: ",
             stringify!(pam_modutil_privs),
@@ -821,9 +592,10 @@
             stringify!(number_of_groups)
         )
     );
+    offset = aligned_offset::<libc::c_int>(offset + ::std::mem::size_of::<libc::c_int>());
     assert_eq!(
         unsafe { ::std::ptr::addr_of!((*ptr).allocated) as usize - ptr as usize },
-        12usize,
+        offset,
         concat!(
             "Offset of field: ",
             stringify!(pam_modutil_privs),
@@ -831,9 +603,10 @@
             stringify!(allocated)
         )
     );
+    offset = aligned_offset::<__gid_t>(offset + ::std::mem::size_of::<libc::c_int>());
     assert_eq!(
         unsafe { ::std::ptr::addr_of!((*ptr).old_gid) as usize - ptr as usize },
-        16usize,
+        offset,
         concat!(
             "Offset of field: ",
             stringify!(pam_modutil_privs),
@@ -841,9 +614,10 @@
             stringify!(old_gid)
         )
     );
+    offset = aligned_offset::<__uid_t>(offset + ::std::mem::size_of::<__gid_t>());
     assert_eq!(
         unsafe { ::std::ptr::addr_of!((*ptr).old_uid) as usize - ptr as usize },
-        20usize,
+        offset,
         concat!(
             "Offset of field: ",
             stringify!(pam_modutil_privs),
@@ -851,9 +625,10 @@
             stringify!(old_uid)
         )
     );
+    offset = aligned_offset::<libc::c_int>(offset + ::std::mem::size_of::<__uid_t>());
     assert_eq!(
         unsafe { ::std::ptr::addr_of!((*ptr).is_dropped) as usize - ptr as usize },
-        24usize,
+        offset,
         concat!(
             "Offset of field: ",
             stringify!(pam_modutil_privs),
@@ -861,100 +636,17 @@
             stringify!(is_dropped)
         )
     );
-}
-extern "C" {
-    pub fn pam_modutil_drop_priv(
-        pamh: *mut pam_handle_t,
-        p: *mut pam_modutil_privs,
-        pw: *const passwd,
-    ) -> libc::c_int;
-}
-extern "C" {
-    pub fn pam_modutil_regain_priv(
-        pamh: *mut pam_handle_t,
-        p: *mut pam_modutil_privs,
-    ) -> libc::c_int;
-}
-pub const pam_modutil_redirect_fd_PAM_MODUTIL_IGNORE_FD: pam_modutil_redirect_fd = 0;
-pub const pam_modutil_redirect_fd_PAM_MODUTIL_PIPE_FD: pam_modutil_redirect_fd = 1;
-pub const pam_modutil_redirect_fd_PAM_MODUTIL_NULL_FD: pam_modutil_redirect_fd = 2;
-pub type pam_modutil_redirect_fd = libc::c_uint;
-extern "C" {
-    pub fn pam_modutil_sanitize_helper_fds(
-        pamh: *mut pam_handle_t,
-        redirect_stdin: pam_modutil_redirect_fd,
-        redirect_stdout: pam_modutil_redirect_fd,
-        redirect_stderr: pam_modutil_redirect_fd,
-    ) -> libc::c_int;
-}
-extern "C" {
-    pub fn pam_modutil_search_key(
-        pamh: *mut pam_handle_t,
-        file_name: *const libc::c_char,
-        key: *const libc::c_char,
-    ) -> *mut libc::c_char;
-}
-pub type __builtin_va_list = [__va_list_tag; 1usize];
-#[repr(C)]
-#[derive(Debug, Copy, Clone)]
-pub struct __va_list_tag {
-    pub gp_offset: libc::c_uint,
-    pub fp_offset: libc::c_uint,
-    pub overflow_arg_area: *mut libc::c_void,
-    pub reg_save_area: *mut libc::c_void,
-}
-#[test]
-fn bindgen_test_layout___va_list_tag() {
-    const UNINIT: ::std::mem::MaybeUninit<__va_list_tag> = ::std::mem::MaybeUninit::uninit();
-    let ptr = UNINIT.as_ptr();
-    assert_eq!(
-        ::std::mem::size_of::<__va_list_tag>(),
-        24usize,
-        concat!("Size of: ", stringify!(__va_list_tag))
-    );
-    assert_eq!(
-        ::std::mem::align_of::<__va_list_tag>(),
-        8usize,
-        concat!("Alignment of ", stringify!(__va_list_tag))
-    );
-    assert_eq!(
-        unsafe { ::std::ptr::addr_of!((*ptr).gp_offset) as usize - ptr as usize },
-        0usize,
-        concat!(
-            "Offset of field: ",
-            stringify!(__va_list_tag),
-            "::",
-            stringify!(gp_offset)
-        )
-    );
-    assert_eq!(
-        unsafe { ::std::ptr::addr_of!((*ptr).fp_offset) as usize - ptr as usize },
-        4usize,
-        concat!(
-            "Offset of field: ",
-            stringify!(__va_list_tag),
-            "::",
-            stringify!(fp_offset)
-        )
-    );
+    offset = aligned_offset::<*mut libc::c_void>(offset + ::std::mem::size_of::<libc::c_int>());
     assert_eq!(
-        unsafe { ::std::ptr::addr_of!((*ptr).overflow_arg_area) as usize - ptr as usize },
-        8usize,
-        concat!(
-            "Offset of field: ",
-            stringify!(__va_list_tag),
-            "::",
-            stringify!(overflow_arg_area)
-        )
-    );
-    assert_eq!(
-        unsafe { ::std::ptr::addr_of!((*ptr).reg_save_area) as usize - ptr as usize },
-        16usize,
-        concat!(
-            "Offset of field: ",
-            stringify!(__va_list_tag),
-            "::",
-            stringify!(reg_save_area)
-        )
+        ::std::mem::size_of::<pam_modutil_privs>(),
+        offset,
+        concat!("Size of: ", stringify!(pam_modutil_privs))
     );
 }
+
+#[cfg(test)]
+fn aligned_offset<T>(offset: usize) -> usize {
+    let offset = offset as isize;
+    let alignment = ::std::mem::align_of::<T>() as isize;
+    (offset + (-offset).rem_euclid(alignment)) as usize
+}
--- a/src/pam/converse.rs
+++ b/src/pam/converse.rs
@@ -25,7 +25,7 @@
     pub fn from_int(val: libc::c_int) -> Option<PamMessageStyle> {
         use PamMessageStyle::*;
 
-        match val as libc::c_uint {
+        match val {
             PAM_PROMPT_ECHO_OFF => Some(PromptEchoOff),
             PAM_PROMPT_ECHO_ON => Some(PromptEchoOn),
             PAM_ERROR_MSG => Some(ErrorMessage),
--- a/src/pam/error.rs
+++ b/src/pam/error.rs
@@ -84,7 +84,7 @@
     pub(super) fn from_int(errno: libc::c_int) -> PamErrorType {
         use PamErrorType::*;
 
-        match errno as libc::c_uint {
+        match errno {
             PAM_SUCCESS => Success,
             PAM_OPEN_ERR => OpenError,
             PAM_SYMBOL_ERR => SymbolError,
